### PR TITLE
feat: add new hook that runs pre resolution

### PR DESCRIPTION
**Pull request by d3lm**
This PR adds a new `preResolve` hook that can be used to run some code before resolution kicks in.

### CODE REVIEW

File: packages/pnpmfile/src/requireHooks.ts [hunk 1]
```diff
@@ -92,3 +105,12 @@ function createReadPackageHookContext (calledFrom: string, prefix: string, hook:
     }),
   }
 }
+
+function createPreResolutionHookLogger (prefix: string): PreResolutionHookLogger {
+  const hook = 'preResolution'
+
+  return {
+    info: (message: string) => hookLogger.info({ message, prefix, hook } as any), // eslint-disable-line
+    warn: (message: string) => hookLogger.warn({ message, prefix, hook } as any), // eslint-disable-line
```
Comment (d3lm) [inline packages/pnpmfile/src/requireHooks.ts:114]: ❓ @zkochan Is it expected that on the `hookLogger` it won't show the yellow warning batch when you call `.warn()`?

Comment (zkochan) [reply packages/pnpmfile/src/requireHooks.ts:114]: Does it show the message at all?

**Review by zkochan — APPROVED** [review]

File: packages/pnpmfile/src/requireHooks.ts (hunk 1 above)
Comment (d3lm) [reply packages/pnpmfile/src/requireHooks.ts:114]: Yes it does, but it's not shown as a warning.

Comment (zkochan) [reply packages/pnpmfile/src/requireHooks.ts:114]: I am not sure why. It is probably how it is implemented in the default-reporter package.

**Pull request merged**